qiime2/sdk/result.py

```diff
@@ -282,6 +286,205 @@ def result(self):
         """
         return self
 
+    def _validate_annotation_support(self):
+        """Checks for the existance of `annotations_dir` on a Result's
+        format class to guard against annotation actions being called on
+        Results with versions < 7.0.
+
+        Raises
+        ------
+        ValueError\n
+            If the Result's format class has no `annotations_dir` and is
+            thus a format version < 7.0.
+
+        """
+        if self._archiver.annotations_dir is None:
+            raise ValueError(
+                'The Artifact or Visualization being used is associated with '
+                'a QIIME 2 archive format of < 7.0. '
+                'Annotation actions are only supported for QIIME 2 archive '
+                'formats of 7.0 and above.'
+            )
+
+    def _read_annotations(self):
+        """
+        Append any existing Annotations to `self._annotations`.
+        Helper method for `add_annotation`, after a given Annotation
+        has been written to disk.
+        """
+        self._annotations = []
```

Comment:
Given the amount of times we iterate over this looking for a given name, this should probably be a dict of {name: annotation}